Tools/cases_generator/generate_cases.py

```diff
@@ -234,7 +234,74 @@ def assign(self, dst: StackEffect, src: StackEffect):
     def cast(self, dst: StackEffect, src: StackEffect) -> str:
         return f"({dst.type or 'PyObject *'})" if src.type != dst.type else ""
 
-INSTRUCTION_FLAGS = ['HAS_ARG', 'HAS_CONST', 'HAS_NAME', 'HAS_JUMP']
+
+class InstructionFlags:
+    """Construct and manipulate instruction flags"""
+
+    INSTRUCTION_FLAGS = ['HAS_ARG', 'HAS_CONST', 'HAS_NAME', 'HAS_JUMP']
+    INSTR_FLAG_SUFFIX = "_FLAG"
+
+    def __init__(self, has_arg: bool, has_const: bool, has_name: bool,
+                 has_jump: bool):
+
+        self.data = {
+            'HAS_ARG': has_arg,
+            'HAS_CONST': has_const,
+            'HAS_NAME': has_name,
+            'HAS_JUMP': has_jump,
+        }
+        assert set(self.data.keys()) == set(self.INSTRUCTION_FLAGS)
+
+    @staticmethod
+    def fromInstruction(instr: "AnyInstruction"):
+        return InstructionFlags(
+            has_arg = variable_used(instr, "oparg"),
+            has_const = variable_used(instr, "FRAME_CO_CONSTS"),
+            has_name = variable_used(instr, "FRAME_CO_NAMES"),
+            has_jump = variable_used(instr, "JUMPBY"),
+        )
+
+    @staticmethod
+    def newEmpty():
+        return InstructionFlags(False, False, False, False)
+
+    def __str__(self) -> str:
+        flags_strs = [f"{name}{self.INSTR_FLAG_SUFFIX}" for i, name in
+                        enumerate(self.INSTRUCTION_FLAGS) if self.is_set(name)]
+        if (len(flags_strs)) == 4: breakpoint()
+        return "0" if not flags_strs else ' | '.join(flags_strs)
+
+    def __eq__(self, other: "InstructionFlags") -> False:
+        return self.bitmap() == other.bitmap()
+
+    def __hash__(self) -> int:
+        return hash(self.bitmap())
+
+    def add(self, other: "InstructionFlags") -> None:
+        for name, value in other.data.items():
+            self.data[name] |= value;
+
+    def bitmap(self) -> int:
+        flags = 0
+        for i, name in enumerate(self.INSTRUCTION_FLAGS):
+            flags |= (1<<i) if self.data[name] else 0
```

Comment:
I think it'd be more useful to have an explicit mapping from flag names to flag values somewhere, rather than using `1<<i` where `i` is the index in `INSTRUCTION_FLAGS`.